dvc/repo/experiments/executor/manager.py

```diff
@@ -0,0 +1,317 @@
+                    logger.error(
+                        "Cancelled before attempting to reproduce experiment "
+                        "'%s'",
+                        rev[:7],
+                    )
+                finally:
+                    executor.cleanup()
+
+        return result
+
+    def _collect_executor(self, executor, exec_result) -> Dict[str, str]:
+        # NOTE: GitPython Repo instances cannot be re-used
+        # after process has received SIGINT or SIGTERM, so we
+        # need this hack to re-instantiate git instances after
+        # checkpoint runs. See:
+        # https://github.com/gitpython-developers/GitPython/issues/427
+        # del self.repo.scm
+
+        results = {}
+
+        def on_diverged(ref: str, checkpoint: bool):
+            ref_info = ExpRefInfo.from_ref(ref)
+            if checkpoint:
+                raise CheckpointExistsError(ref_info.name)
+            raise ExperimentExistsError(ref_info.name)
+
+        for ref in executor.fetch_exps(
+            self.scm,
+            executor.git_url,
+            force=exec_result.force,
+            on_diverged=on_diverged,
+        ):
+            exp_rev = self.scm.get_ref(ref)
+            if exp_rev:
+                logger.debug("Collected experiment '%s'.", exp_rev[:7])
+                results[exp_rev] = exec_result.exp_hash
+
+        return results
+
+
+class TempDirExecutorManager(BaseExecutorManager):
+    EXECUTOR_CLS = TempDirExecutor
+
+
+class WorkspaceExecutorManager(BaseExecutorManager):
+    EXECUTOR_CLS = WorkspaceExecutor
+
+    @classmethod
+    def from_stash_entries(
+        cls,
+        scm: "Git",
+        wdir: str,
+        repo: "Repo",
+        to_run: Dict[str, ExpStashEntry],
+        **kwargs,
+    ):
+        manager = cls(scm, wdir)
+        try:
+            assert len(to_run) == 1
+            for stash_rev, entry in to_run.items():
+                scm.set_ref(EXEC_HEAD, entry.head_rev)
+                scm.set_ref(EXEC_MERGE, stash_rev)
+                scm.set_ref(EXEC_BASELINE, entry.baseline_rev)
+
+                executor = cls.EXECUTOR_CLS.from_stash_entry(
+                    repo,
+                    stash_rev,
+                    entry,
+                    **kwargs,
+                )
+                manager.enqueue(stash_rev, executor)
+        finally:
+            for ref in (EXEC_MERGE,):
+                scm.remove_ref(ref)
+        return manager
+
+    def _collect_executor(self, executor, exec_result) -> Dict[str, str]:
+        results = {}
+        exp_rev = self.scm.get_ref(EXEC_BRANCH)
+        if exp_rev:
+            logger.debug("Collected experiment '%s'.", exp_rev[:7])
+            results[exp_rev] = exec_result.exp_hash
+        return results
+
+    def exec_queue(self, jobs: Optional[int] = 1):
+        """Run a single WorkspaceExecutor.
+
+        Workspace execution is done within the main DVC process
+        (rather than in multiprocessing context)
+        """
```

Comment:
In theory we could get rid of this entirely and just use `WorkspaceExecutor` inside the generic base multiprocessing context, but for now keeping the workspace specific behavior at least makes it easier to test exp reproduction (since we don't have to deal with testing function calls from inside multiprocessing child workers).